### ISSUE TITLE
Items put on Queue in thread in child process sometimes not seen by parent process

### ISSUE THREAD

**Issue by f5efdf1f-0eca-4f50-ad08-3aa59665a7f8** [python/cpython#85780]
BPO | [41614](https://bugs.python.org/issue41614)
--- | :---
Nosy | @alexmojaki
Files | <li>[queue_bug.py](https://bugs.python.org/file49422/queue_bug.py "Uploaded as text/plain at 2020-08-22.11:42:14 by @alexmojaki"): A script you can run to observe this bug, with extensive comments</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-08-22.11:42:15.115>
labels = ['3.7', '3.8', '3.10']
title = 'Items put on Queue in thread in child process sometimes not seen by parent process'
updated_at = <Date 2020-08-22.11:42:15.115>
user = 'https://github.com/alexmojaki'
```

bugs.python.org fields:
```python
activity = <Date 2020-08-22.11:42:15.115>
actor = 'alexmojaki'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2020-08-22.11:42:15.115>
creator = 'alexmojaki'
dependencies = []
files = ['49422']
hgrepos = []
issue_num = 41614
keywords = []
message_count = 1.0
messages = ['375795']
nosy_count = 1.0
nosy_names = ['alexmojaki']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue41614'
versions = ['Python 3.6', 'Python 3.7', 'Python 3.8', 'Python 3.10']
```

</p></details>

**Comment by duaneg** [comment]
FWIW I was not able to reproduce this with current HEAD

**Comment by picnixz** [comment]
> FWIW I was not able to reproduce this with current HEAD.

Could it be because the *default* start method was changed?

**Comment by duaneg** [comment]
> Could it be because the _default_ start method was changed? For Linux/macOS, have you tried using the `fork` start method explicitly for `multiprocessing`? because now, we're using `forkserver` which may not suffer from this issue.

A good thought but in this case, no: I have bisected the fix to #83856. Although according to the discussion in the PR the `spawn` method would probably not have been susceptible to this bug, so using that would possibly be a viable work-around for users affected who need to stay on older versions.

**Comment by picnixz** [comment]
3.13 and 3.14 were fixed by https://github.com/python/cpython/pull/114279, but could 3.12 still suffer from this? if so, we could possibly fix it but the last bugfix is near.

**Comment by duaneg** [comment]
> 3.13 and 3.14 were fixed by [#114279](https://github.com/python/cpython/pull/114279), but could 3.12 still suffer from this? if so, we could possibly fix it but the last bugfix is near.

Yes, this does affect 3.12. I just tested and confirmed it affects all three methods in that version. FTR I have also tested all three methods in HEAD and confirmed all are working correctly.

Whether we should back-port the fix for a 3.12 bugfix release I will leave to those qualified to decide :wink:

**Comment by picnixz** [comment]
cc @gpshead
